python/lsst/daf/butler/column_spec.py

```diff
@@ -0,0 +1,310 @@
+
+
+@final
+class HashColumnSpec(ColumnSpec):
+    """Description of a hash digest."""
+
+    pytype: ClassVar[type] = bytes
+
+    type: Literal["hash"] = "hash"
+
+    nbytes: int
+    """Number of bytes for the hash."""
+
+    def to_sql_spec(self, **kwargs: Any) -> ddl.FieldSpec:
+        # Docstring inherited.
+        return super().to_sql_spec(nbytes=self.nbytes, **kwargs)
+
+    def to_arrow(self) -> arrow_utils.ToArrow:
+        # Docstring inherited.
+        assert self.nullable is not None, "nullable=None should be resolved by validators"
+        return arrow_utils.ToArrow.for_primitive(
+            self.name,
+            # The size for Arrow binary columns is a fixed size, not a maximum
+            # as in SQL, so we use a variable-size column.
+            pa.binary(),
+            nullable=self.nullable,
+        )
+
+
+@final
+class FloatColumnSpec(ColumnSpec):
+    """Description of a float column."""
+
+    pytype: ClassVar[type] = float
+
+    type: Literal["float"] = "float"
+
+    def to_arrow(self) -> arrow_utils.ToArrow:
+        # Docstring inherited.
+        assert self.nullable is not None, "nullable=None should be resolved by validators"
+        return arrow_utils.ToArrow.for_primitive(self.name, pa.float64(), nullable=self.nullable)
+
+
+@final
+class BoolColumnSpec(ColumnSpec):
+    """Description of a bool column."""
+
+    pytype: ClassVar[type] = bool
+
+    type: Literal["bool"] = "bool"
+
+    def to_arrow(self) -> arrow_utils.ToArrow:
+        # Docstring inherited.
+        assert self.nullable is not None, "nullable=None should be resolved by validators"
+        return arrow_utils.ToArrow.for_primitive(self.name, pa.bool_(), nullable=self.nullable)
+
+
+@final
+class RegionColumnSpec(ColumnSpec):
+    """Description of a region column."""
+
+    name: str = "region"
+
+    pytype: ClassVar[type] = Region
+
+    type: Literal["region"] = "region"
+
+    nbytes: int = 2048
+    """Number of bytes for the encoded region."""
+
+    def to_arrow(self) -> arrow_utils.ToArrow:
+        # Docstring inherited.
+        assert self.nullable is not None, "nullable=None should be resolved by validators"
+        return arrow_utils.ToArrow.for_region(self.name, nullable=self.nullable)
+
+
+@final
+class TimespanColumnSpec(ColumnSpec):
+    """Description of a timespan column."""
+
+    name: str = "timespan"
+
+    pytype: ClassVar[type] = Timespan
+
+    type: Literal["timespan"] = "timespan"
+
+    def to_arrow(self) -> arrow_utils.ToArrow:
+        # Docstring inherited.
+        assert self.nullable is not None, "nullable=None should be resolved by validators"
+        return arrow_utils.ToArrow.for_timespan(self.name, nullable=self.nullable)
+
+
+AnyColumnSpec = Union[
+    IntColumnSpec,
+    StringColumnSpec,
+    HashColumnSpec,
+    FloatColumnSpec,
+    BoolColumnSpec,
+    RegionColumnSpec,
+    TimespanColumnSpec,
+]
```

Comment:
Not totally clear to me.  Per the documentation it definitely works if the discriminators are different between the outer one and the inner one, but I don't think that's what you're asking about: https://docs.pydantic.dev/latest/concepts/unions/#nested-discriminated-unions

I haven't tried it but I don't see any reason something like:
```
A = Annotated[Union[...], Field(discriminator='type')]
B = Annotated[Union[...], Field(discriminator='type')]
Combined = Union[A, B]
```
wouldn't work, since in the combined Union it's checking each member separately.  You'd get two separate error messages if you tried to include something that wasn't in either union but that's probably fine.

What's less clear (and is possibly documented to NOT work but it's slightly ambiguous):
```
# ????
Combined = Annotated[Union[A, B], Field(discriminator='type')]
```